docs/polymorphic-table-types.md

```diff
@@ -0,0 +1,198 @@
+# Polymorphic Table Types
+
+## Summary
+
+We propose implementing polymorphic tables of the form `<T>{ [F<T>] : G<T> }` analogous to already-existing polymorphic functions `<T>(F<T>) -> G<T>`.
+
+## Motivation
+
+The goal of this RFC is furthering type-safety and type-representation, especially in niche or performance-critical applications. Polymorphic Functions exist already in the form `<T>(F<T>) -> G<T>`, where `T` is inferred at the call site, with similar applications.
+
+At a type level, luau functions and tables are homomorphic as type mappings. One can view a table as a function mapping its keys to its values, or a function as a table with inputs as keys and outputs as values. This change resolves the hole left behind implementing polymorphic functions while excluding polymorphic tables, forcing type-safe applications to prefer functions over tables as maps.
```

Comment:
I'm not really sure that I agree that functions and tables are homomorphic. The term homomorphic refers to the idea that there is a _structure-preserving_ mapping between the underlying sets of two algebras. I don't believe this is the case for functions and tables. Tables are a finite mapping, they intrinsically behave differently as a result. As far as I can tell, there does not exist any values for your proposed polymorphic table types where the codomain is not an optional type since returning `nil` is the only thing that can allow you to act as if tables have an infinite domain.

This kind of thing matters in practice because, ultimately, it's not a design choice we've made to force people to use functions over tables, but rather an innate property of functions that they are well-behaved mappings over potentially-infinite domains.